src/shared/dissect-image.c

```diff
@@ -1125,10 +1120,38 @@ int dissect_image(
                 }
         }
 
-        blkid_free_probe(b);
-        b = NULL;
+        return 0;
+}
+#endif
 
-        r = dissected_image_probe_filesystem(m);
+int dissect_image_file(
+                const char *path,
+                const VeritySettings *verity,
+                const MountOptions *mount_options,
+                DissectImageFlags flags,
+                DissectedImage **ret) {
+
+#if HAVE_BLKID
+        _cleanup_(dissected_image_unrefp) DissectedImage *m = NULL;
+        _cleanup_close_ int fd = -1;
+        int r;
+
+        assert(path);
+        assert(ret);
+
+        fd = open(path, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
+        if (fd < 0)
+                return -errno;
+
+        r = fd_verify_regular(fd);
+        if (r < 0)
+                return r;
+
+        r = dissected_image_new(path, &m);
+        if (r < 0)
+                return r;
+
+        r = dissect_image(fd, path, verity, mount_options, flags & ~DISSECT_IMAGE_BLOCK_DEVICE, m);
```

Comment:
why mask off the DISSECT_IMAGE_BLOCK_DEVICE flag? shouldn't we rather generate an early error here if it is set? i.e.

```c
if (FLAGS_SET(flags, DISSECT_IMAGE_BLOCK_DEVICE))
        return -EINVAL;
```

i guess assert() would be ok to, since it's pretty much a programming error?